lib/Dialect/FIRRTL/Transforms/CheckInit.cpp

```diff
@@ -0,0 +1,318 @@
+  void emitInitError(Value decl, size_t fieldID);
+
+public:
+  void runOnOperation() override;
+};
+} // end anonymous namespace
+
+static bool checkBit(BitVector &vals, size_t idx) {
+  if (vals.size() <= idx)
+    return false;
+  return vals[idx];
+}
+
+static std::string nameForField(Type t, size_t fieldID) {
+  LLVM_DEBUG(llvm::errs() << "N: " << t << " @" << fieldID << "\n";);
+  return FIRRTLTypeSwitch<Type, std::string>(t)
+      .Case<BundleType, OpenBundleType>([fieldID](auto bundle) {
+        auto idx = bundle.getIndexForFieldID(fieldID);
+        auto fid = bundle.getFieldID(idx);
+        if (fieldID == fid)
+          return bundle.getElementName(idx).str();
+        return bundle.getElementName(idx).str() + "." +
+               nameForField(bundle.getElementType(idx), fieldID - fid);
+      })
+      .Case<FVectorType, OpenVectorType>([fieldID](auto vec) {
+        auto idx = vec.getIndexForFieldID(fieldID);
+        auto fid = vec.getFieldID(idx);
+        if (fieldID == fid)
+          return "[" + std::to_string(idx) + "]";
+        return "[" + std::to_string(idx) + "]." +
+               nameForField(vec.getElementType(), fieldID - fid);
+      })
+      .Default([](auto) { return "INVALID"; });
+}
+
+static StringRef getPortName(BlockArgument b) {
+  FModuleOp mod = cast<FModuleOp>(b.getOwner()->getParentOp());
+  return mod.getPort(b.getArgNumber()).getName();
+}
+
+static StringRef getMemPortName(Value v) {
+  MemOp mem = cast<MemOp>(v.getDefiningOp());
+  return mem.getPortName(cast<OpResult>(v).getResultNumber());
+}
+
+static StringRef getCMemPortName(Value v) {
+  chirrtl::MemoryPortOp mem = cast<chirrtl::MemoryPortOp>(v.getDefiningOp());
+  return mem.getName();
+}
+
+static StringRef getInstPortName(Value v) {
+  InstanceOp inst = cast<InstanceOp>(v.getDefiningOp());
+  return inst.getPortName(cast<OpResult>(v).getResultNumber());
+}
+
+void CheckInitPass::emitInitError(Value decl, size_t fieldID) {
+
+  InFlightDiagnostic err = [&]() {
+    if (auto ba = dyn_cast<BlockArgument>(decl))
+      return getOperation()->emitError("Port `") << getPortName(ba) << "`";
+    if (isa<InstanceOp>(decl.getDefiningOp()))
+      return decl.getDefiningOp()->emitError("Instance input `")
+             << getInstPortName(decl) << "`";
+    if (isa<chirrtl::MemoryPortOp>(decl.getDefiningOp()))
+      return decl.getDefiningOp()->emitError("Memory port `")
+             << getCMemPortName(decl) << "`";
+    if (isa<MemOp>(decl.getDefiningOp()))
+      return decl.getDefiningOp()->emitError("Memory port `")
+             << getMemPortName(decl) << "`";
+    return decl.getDefiningOp()->emitError("Wire");
+  }();
+  if (fieldID == 0) {
+    err << " is not initialized.";
+  } else {
+    err << " is not fully initialized at field `"
+        << nameForField(decl.getType(), fieldID) << "`.";
+  }
+  failed = true;
+}
+
+// Check (recursively) that decl at fieldID, whose type at fieldID is t, has a
+// bit set in vals
```

Comment:
```suggestion
// bit set in vals.
```
Usually we could start the next sentence on the same line, I think?